DS8K rescan improvement (UB-1103)

hfeish opened the pull request:
Installation of Ubiquity fails on DS8k because rescan-scsi-bus.sh hit a LUN0 issue, to avoid this we use "rescan-scsi-bus.sh -a" if the previous rescan "rescan-scsi-bus.sh -r" fail, and since "rescan-scsi-bus.sh -a" is slowly than "rescan-scsi-bus.sh -r", we still use the "rescan-scsi-bus.sh -r" as default rescan

Below is the procedure:

    1. RescanAll with "rescan-scsi-bus.sh -r"
    2 .Discover device by WWN
              If cannot find WWN, judge that if this is FC env, use "rescan-scsi-bus.sh -a" to rescan for the second time, wwn will show in the "multipath -ll",
              If can find WWN, doesn't need second rescan

Run sanity passed, will run with XAVI



Signed-off-by: feihuang <feihuang@feihuangs-mbp.cn.ibm.com>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ibm/ubiquity/204)
<!-- Reviewable:end -->

hfeish commented:
@tzurE , please help to review, thanks!

ranhrl commented:
Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---

*[remote/mounter/scbe.go, line 58 at r1](https://beta.reviewable.io/reviews/ibm/ubiquity/204#-L9dZ17D4nYA5eV7d6wV:-L9dZ17D4nYA5eV7d6wW:b-tcl4hy) ([raw file](https://github.com/ibm/ubiquity/blob/79c25c335ba6920af56c413102a6ef73187898a2/remote/mounter/scbe.go#L58)):*
> ```Go
> 		// Known issue with rescan-scsi-bus.sh, LUN0 must be the first mapped logical unit, otherwise it will not be able to scan any other logical unit
> 		// For DS8k, it only support FC and without LUN0 mapped default, so we need to do "rescan-scsi-bus.sh -a" to discover the LUN0
> 		if s.config.SkipRescanISCSI {
> ```

This name "SkipRescanISCSI" is a bad way of saying "ConfigurationIsISCSI" or "ConfigurationIsFC". I already discussed it once with Shay, I am not sure if there is a ticket to change it to a more readable name.

---

*[remote/mounter/scbe.go, line 60 at r1](https://beta.reviewable.io/reviews/ibm/ubiquity/204#-L9dZkXP9QrD86_Om7qf:-L9dZkXP9QrD86_Om7qg:b-4tmocj) ([raw file](https://github.com/ibm/ubiquity/blob/79c25c335ba6920af56c413102a6ef73187898a2/remote/mounter/scbe.go#L60)):*
> ```Go
> 		if s.config.SkipRescanISCSI {
> 			if err := s.blockDeviceMounterUtils.RescanAllTargets(!s.config.SkipRescanISCSI, volumeWWN); err != nil {
> 				return "", s.logger.ErrorRet(err, "RescanAll failed again")
> ```

Please be more accurate with the message of the error. "RescanAll failed with error code {} for device {volumeWWN}"

---

*[remote/mounter/block_device_utils/rescan.go, line 27 at r1](https://beta.reviewable.io/reviews/ibm/ubiquity/204#-L9dd46u0lQt0A7g8JV3:-L9dd46u0lQt0A7g8JV4:br3zuzm) ([raw file](https://github.com/ibm/ubiquity/blob/79c25c335ba6920af56c413102a6ef73187898a2/remote/mounter/block_device_utils/rescan.go#L27)):*
> ```Go
> 	defer b.logger.Trace(logs.DEBUG)()
> 
> 	switch protocol {
> ```

protocol value could be either FC, ISCSI or both, I think we should refactor to reflect it. It is not clear what SCSI means if you have FC and ISCSI. 

---

*[remote/mounter/block_device_utils/rescan.go, line 76 at r1](https://beta.reviewable.io/reviews/ibm/ubiquity/204#-L9dcl1e7BOY0cP42wYA:-L9dcl1e7BOY0cP42wYB:bcthmnp) ([raw file](https://github.com/ibm/ubiquity/blob/79c25c335ba6920af56c413102a6ef73187898a2/remote/mounter/block_device_utils/rescan.go#L76)):*
> ```Go
> 	commands := []string{"rescan-scsi-bus", "rescan-scsi-bus.sh"}
> 	rescanCmd := ""
> 	for _, cmd := range commands {
> ```

Code duplication, you should have a separate method to get the correct rescan command and use it for both.

---

*[remote/mounter/block_device_utils/rescan.go, line 85 at r1](https://beta.reviewable.io/reviews/ibm/ubiquity/204#-L9dcNZ32M95tFBH-p5X:-L9dcNZ32M95tFBH-p5Y:b-u0kt69) ([raw file](https://github.com/ibm/ubiquity/blob/79c25c335ba6920af56c413102a6ef73187898a2/remote/mounter/block_device_utils/rescan.go#L85)):*
> ```Go
> 		return b.logger.ErrorRet(&commandNotFoundError{commands[0], errors.New("")}, "failed")
> 	}
> 	args := []string{"-a"} // Only for the Lun0 rescan on DS8k
> ```

the comment should state rescan all targets if lun was not find in order to find a lun0 on a new target.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

tzurE commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[remote/mounter/scbe.go, line 58 at r1](https://beta.reviewable.io/reviews/ibm/ubiquity/204#-L9dZ17D4nYA5eV7d6wV:-L9e-dlzEtrJYO9S1T3P:b32v57) ([raw file](https://github.com/ibm/ubiquity/blob/79c25c335ba6920af56c413102a6ef73187898a2/remote/mounter/scbe.go#L58)):*
<details><summary><i>Previously, ranhrl (Ran Harel) wrote…</i></summary><blockquote>

This name "SkipRescanISCSI" is a bad way of saying "ConfigurationIsISCSI" or "ConfigurationIsFC". I already discussed it once with Shay, I am not sure if there is a ticket to change it to a more readable name.
</blockquote></details>

+1

---

*[remote/mounter/block_device_utils/rescan.go, line 76 at r1](https://beta.reviewable.io/reviews/ibm/ubiquity/204#-L9dcl1e7BOY0cP42wYA:-L9e1mi53BZ-iwsBEl8s:b32v57) ([raw file](https://github.com/ibm/ubiquity/blob/79c25c335ba6920af56c413102a6ef73187898a2/remote/mounter/block_device_utils/rescan.go#L76)):*
<details><summary><i>Previously, ranhrl (Ran Harel) wrote…</i></summary><blockquote>

Code duplication, you should have a separate method to get the correct rescan command and use it for both.
</blockquote></details>

+1

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

hfeish commented:
Review status: 0 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[remote/mounter/scbe.go, line 58 at r1](https://beta.reviewable.io/reviews/ibm/ubiquity/204#-L9dZ17D4nYA5eV7d6wV:-L9hNj0r1lfD1I45lpZ3:bvzuno) ([raw file](https://github.com/ibm/ubiquity/blob/79c25c335ba6920af56c413102a6ef73187898a2/remote/mounter/scbe.go#L58)):*
<details><summary><i>Previously, tzurE wrote…</i></summary><blockquote>

+1
</blockquote></details>

I don't find the ticket in Jira, do we need to open a new ticket to handle this or fix this in this PR? If we need to fix now, any details about the new name?

---

*[remote/mounter/scbe.go, line 60 at r1](https://beta.reviewable.io/reviews/ibm/ubiquity/204#-L9dZkXP9QrD86_Om7qf:-L9hQh-i9S_inRtKLytu:b-mjjxwg) ([raw file](https://github.com/ibm/ubiquity/blob/79c25c335ba6920af56c413102a6ef73187898a2/remote/mounter/scbe.go#L60)):*
<details><summary><i>Previously, ranhrl (Ran Harel) wrote…</i></summary><blockquote>

Please be more accurate with the message of the error. "RescanAll failed with error code {} for device {volumeWWN}"
</blockquote></details>

Changed to s.logger.ErrorRet(err, "RescanAll Targets failed", logs.Args{{"volumeWWN", volumeWWN}})

---

*[remote/mounter/block_device_utils/rescan.go, line 27 at r1](https://beta.reviewable.io/reviews/ibm/ubiquity/204#-L9dd46u0lQt0A7g8JV3:-L9hQvsW3py9DGo24B5b:b-pj5xop) ([raw file](https://github.com/ibm/ubiquity/blob/79c25c335ba6920af56c413102a6ef73187898a2/remote/mounter/block_device_utils/rescan.go#L27)):*
<details><summary><i>Previously, ranhrl (Ran Harel) wrote…</i></summary><blockquote>

protocol value could be either FC, ISCSI or both, I think we should refactor to reflect it. It is not clear what SCSI means if you have FC and ISCSI. 
</blockquote></details>

Changed to BOTH, ISCSI, FC

---

*[remote/mounter/block_device_utils/rescan.go, line 76 at r1](https://beta.reviewable.io/reviews/ibm/ubiquity/204#-L9dcl1e7BOY0cP42wYA:-L9hR1Cq8yRgGM7Q-vyP:b-rwxi41) ([raw file](https://github.com/ibm/ubiquity/blob/79c25c335ba6920af56c413102a6ef73187898a2/remote/mounter/block_device_utils/rescan.go#L76)):*
<details><summary><i>Previously, tzurE wrote…</i></summary><blockquote>

+1
</blockquote></details>

Changed to pass "-a"  or "-r" as a string arg to the RescanSCSI

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

ranhrl commented:
Review status: 0 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[remote/mounter/scbe.go, line 58 at r1](https://beta.reviewable.io/reviews/ibm/ubiquity/204#-L9dZ17D4nYA5eV7d6wV:-LAIR-oiAsLRfD6hXbD8:b-hjfn5c) ([raw file](https://github.com/ibm/ubiquity/blob/79c25c335ba6920af56c413102a6ef73187898a2/remote/mounter/scbe.go#L58)):*
<details><summary><i>Previously, hfeish (Fei Huang) wrote…</i></summary><blockquote>

I don't find the ticket in Jira, do we need to open a new ticket to handle this or fix this in this PR? If we need to fix now, any details about the new name?
</blockquote></details>

You can open a new ticket and discuss it with Shay when he is back

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

hfeish commented:
Review status: 0 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[remote/mounter/scbe.go, line 58 at r1](https://beta.reviewable.io/reviews/ibm/ubiquity/204#-L9dZ17D4nYA5eV7d6wV:-LALXe5Z1YQHUKJYtFyV:blabla) ([raw file](https://github.com/ibm/ubiquity/blob/79c25c335ba6920af56c413102a6ef73187898a2/remote/mounter/scbe.go#L58)):*
<details><summary><i>Previously, ranhrl (Ran Harel) wrote…</i></summary><blockquote>

You can open a new ticket and discuss it with Shay when he is back
</blockquote></details>

Open UB-1187

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

shay-berman commented:
Review status: 0 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[a discussion](https://reviewable.io/reviews/ibm/ubiquity/204#-LBtq89UERseZqve0TTf:-LBtq89UERseZqve0TTg:b-madndv) (no related file):*
Is this fix is mandatory?   I don't like that we doing additional rescan if we don't find the device. Lets talk about it offline.

---


*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

shay-berman commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[a discussion](https://reviewable.io/reviews/ibm/ubiquity/204#-LBtq89UERseZqve0TTf:-LDjUUXIEe1KmXMktGbz:b-mmh4vl) (no related file):*
<details><summary><i>Previously, ranhrl (Ran Harel) wrote…</i></summary><blockquote>

What is the alternative, Shay?
</blockquote></details>


if rescan -a is only for ds8k use case, then I would like to make sure it will run only for ds8k use case. For example, you can identify if the wwn that you want to discover is for ds8k, if so then you can add this flow (if the first discover failed then run another one with rescan -a. but only for ds8k luns).  
In xiv you have the ability to identify if the wwn is xiv or a9000 system, maybe its the same for ds8k.

---


*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

hfeish commented:
@shay-berman , please help to review, thanks

hfeish commented:
@shay-berman , please help to review

beckmani commented:
*[local/scbe/scbe.go, line 446 at r3](https://reviewable.io/reviews/ibm/ubiquity/204#-LFkXybfEtydYHhz3fiZ:-LFkXybfEtydYHhz3fi_:b-55thv4) ([raw file](https://github.com/ibm/ubiquity/blob/e87f9ea7ff6e2a0666c2393aacb0d71bb0d52a99/local/scbe/scbe.go#L446)):*
> ```Go
> 
> 	//Save the lun number in the map, if get the same wwn, then update the lunNumber with latest
> 	lunNumberDict[existingVolume.WWN] = scbeResponseMap.LunNumber
> ```

I suggest that you log the case where the lunNumber in the dict is different from the current lun number

---


*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

beckmani commented:
*[remote/mounter/scbe.go, line 46 at r3](https://reviewable.io/reviews/ibm/ubiquity/204#-LFkYUhz-ddJukfb_IMY:-LFkYUhz-ddJukfb_IMZ:bytea1d) ([raw file](https://github.com/ibm/ubiquity/blob/e87f9ea7ff6e2a0666c2393aacb0d71bb0d52a99/remote/mounter/scbe.go#L46)):*
> ```Go
> func NewTestScbeMounter(scbeRemoteConfig resources.ScbeRemoteConfig, blockDeviceMounterUtils block_device_mounter_utils.BlockDeviceMounterUtils) resources.Mounter {
> 	return &scbeMounter{
> 		logger:                  logs.GetLogger(),
> ```

Too many spaces.

---


*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

beckmani commented:
*[remote/mounter/scbe.go, line 66 at r3](https://reviewable.io/reviews/ibm/ubiquity/204#-LFkYdIXB_v4fflJbGak:-LFkYdIXB_v4fflJbGal:bunfkvg) ([raw file](https://github.com/ibm/ubiquity/blob/e87f9ea7ff6e2a0666c2393aacb0d71bb0d52a99/remote/mounter/scbe.go#L66)):*
> ```Go
> 	if err != nil {
> 		// Known issue with rescan-scsi-bus.sh, LUN0 must be the first mapped logical unit, otherwise it will not be able to scan any other logical unit
> 		// For DS8k, it only support FC and without LUN0 mapped default, so we need to do "rescan-scsi-bus.sh -a" to discover the LUN0
> ```

Please add a reference url to RH known issue.

---


*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

beckmani commented:
*[remote/mounter/scbe_test.go, line 6 at r3](https://reviewable.io/reviews/ibm/ubiquity/204#-LFkZU4-E0BjmzYiPTWo:-LFkZU4-E0BjmzYiPTWp:b2e6o7d) ([raw file](https://github.com/ibm/ubiquity/blob/e87f9ea7ff6e2a0666c2393aacb0d71bb0d52a99/remote/mounter/scbe_test.go#L6)):*
> ```Go
> 	. "github.com/IBM/ubiquity/remote/mounter"
> 	. "github.com/IBM/ubiquity/resources"
> 	"github.com/IBM/ubiquity/fakes"
> ```

Nit: Alignment

---


*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

beckmani commented:
*[remote/mounter/block_device_mounter_utils/block_device_mounter_utils.go, line 200 at r3](https://reviewable.io/reviews/ibm/ubiquity/204#-LFk_N8WBnXLsyn0XxYc:-LFk_N8WBnXLsyn0XxYd:bvljn75) ([raw file](https://github.com/ibm/ubiquity/blob/e87f9ea7ff6e2a0666c2393aacb0d71bb0d52a99/remote/mounter/block_device_mounter_utils/block_device_mounter_utils.go#L200)):*
> ```Go
> 		}
> 		b.logger.Debug("rescanLock.TryLock failed", logs.Args{{"error", err}})
> 		time.Sleep(time.Duration(500*time.Millisecond))
> ```

Move 500 to be constant

---


*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

beckmani commented:
*[remote/mounter/scbe_test.go, line 24 at r3](https://reviewable.io/reviews/ibm/ubiquity/204#-LFk_t4D2PpsaJkA599w:-LFk_t4D2PpsaJkA599x:b-j3gh96) ([raw file](https://github.com/ibm/ubiquity/blob/e87f9ea7ff6e2a0666c2393aacb0d71bb0d52a99/remote/mounter/scbe_test.go#L24)):*
> ```Go
> 
> 		mountRequestForDS8kLun0 = MountRequest{Mountpoint:"test_mountpointDS8k", VolumeConfig:map[string]interface{}{"Name":"u_vol","PhysicalCapacity":2040,
> 		"Profile":"gold", "StorageType":"2107", "UsedCapacity":2040, "Wwn":"wwn", "attach-to":"xnode1",
> ```

Use const for PhysicalCapacity, StorageType, Profile, and  UsedCapacity (Here and below)

---


*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

beckmani commented:
Reviewed 2 of 7 files at r1, 5 of 6 files at r2, 6 of 6 files at r3.
Review status: all files reviewed, 11 unresolved discussions (waiting on @ranhrl, @tzurE, @hfeish, and @alonm)

---



*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

hfeish commented:
Review status: 10 of 13 files reviewed, 11 unresolved discussions (waiting on @beckmani, @ranhrl, @tzurE, @hfeish, and @alonm)

---

*[local/scbe/scbe.go, line 446 at r3](https://reviewable.io/reviews/ibm/ubiquity/204#-LFkXybfEtydYHhz3fiZ:-LFotaYF00Bqmifwpo0a:bb6un1j) ([raw file](https://github.com/ibm/ubiquity/blob/e87f9ea7ff6e2a0666c2393aacb0d71bb0d52a99/local/scbe/scbe.go#L446)):*
<details><summary><i>Previously, beckmani (Isaac Beckman) wrote…</i></summary><blockquote>

I suggest that you log the case where the lunNumber in the dict is different from the current lun number
</blockquote></details>

Done

---

*[remote/mounter/scbe.go, line 46 at r3](https://reviewable.io/reviews/ibm/ubiquity/204#-LFkYUhz-ddJukfb_IMY:-LFotetp-h1fSOusy56K:bdmeep1) ([raw file](https://github.com/ibm/ubiquity/blob/e87f9ea7ff6e2a0666c2393aacb0d71bb0d52a99/remote/mounter/scbe.go#L46)):*
<details><summary><i>Previously, beckmani (Isaac Beckman) wrote…</i></summary><blockquote>

Too many spaces.
</blockquote></details>

Run with gofmt

---

*[remote/mounter/scbe.go, line 66 at r3](https://reviewable.io/reviews/ibm/ubiquity/204#-LFkYdIXB_v4fflJbGak:-LFotiirEXtUP7jOhH7N:bh1wg1x) ([raw file](https://github.com/ibm/ubiquity/blob/e87f9ea7ff6e2a0666c2393aacb0d71bb0d52a99/remote/mounter/scbe.go#L66)):*
<details><summary><i>Previously, beckmani (Isaac Beckman) wrote…</i></summary><blockquote>

Please add a reference url to RH known issue.
</blockquote></details>

Add url here

---

*[remote/mounter/scbe_test.go, line 6 at r3](https://reviewable.io/reviews/ibm/ubiquity/204#-LFkZU4-E0BjmzYiPTWo:-LFotojnBW8grU0uOPLo:bu6yll1) ([raw file](https://github.com/ibm/ubiquity/blob/e87f9ea7ff6e2a0666c2393aacb0d71bb0d52a99/remote/mounter/scbe_test.go#L6)):*
<details><summary><i>Previously, beckmani (Isaac Beckman) wrote…</i></summary><blockquote>

Nit: Alignment
</blockquote></details>

run with gofmt

---

*[remote/mounter/scbe_test.go, line 24 at r3](https://reviewable.io/reviews/ibm/ubiquity/204#-LFk_t4D2PpsaJkA599w:-LFotsyCE3cmJbXqkhIU:bb7f2mf) ([raw file](https://github.com/ibm/ubiquity/blob/e87f9ea7ff6e2a0666c2393aacb0d71bb0d52a99/remote/mounter/scbe_test.go#L24)):*
<details><summary><i>Previously, beckmani (Isaac Beckman) wrote…</i></summary><blockquote>

Use const for PhysicalCapacity, StorageType, Profile, and  UsedCapacity (Here and below)
</blockquote></details>

done

---

*[remote/mounter/block_device_mounter_utils/block_device_mounter_utils.go, line 200 at r3](https://reviewable.io/reviews/ibm/ubiquity/204#-LFk_N8WBnXLsyn0XxYc:-LFotvF6BVGUPUk60pnT:bkfpq2f) ([raw file](https://github.com/ibm/ubiquity/blob/e87f9ea7ff6e2a0666c2393aacb0d71bb0d52a99/remote/mounter/block_device_mounter_utils/block_device_mounter_utils.go#L200)):*
<details><summary><i>Previously, beckmani (Isaac Beckman) wrote…</i></summary><blockquote>

Move 500 to be constant
</blockquote></details>

Check that all the para here not use constant, so keep constent

---


*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

shay-berman commented:
Review status: 10 of 13 files reviewed, 25 unresolved discussions (waiting on @beckmani, @ranhrl, @tzurE, @hfeish, and @alonm)

---

*[Dockerfile, line 10 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq-enzBExMqiv6VhlU:-LFq-eo1-SGJ4ZcKRRnH:bieyocr) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/Dockerfile#L10)):*
> ```Dockerfile
> 
> FROM alpine:3.7
> RUN apk --no-cache add ca-certificates=20171114-r0 openssl=1.0.2o-r0
> ```

??
how its related to the PR?

---

*[local/scbe/scbe.go, line 55 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq0lHr4VBUMU9yQeLz:-LFq0lHr4VBUMU9yQeM-:b-ktb16o) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/local/scbe/scbe.go#L55)):*
> ```Go
> var (
> 	SupportedFSTypes = []string{"ext4", "xfs"}
> 	lunNumberDict = make(map[string]int)
> ```

what is this?  be careful ubiquity is thread save 

---

*[local/scbe/scbe.go, line 388 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq0FJDAUfhkT3W01nH:-LFq0FJDAUfhkT3W01nI:bfm8zxk) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/local/scbe/scbe.go#L388)):*
> ```Go
> 	volConfig[resources.ScbeKeyVolAttachToHost] = hostAttach
> 
> 	// Mounter mount will use this extra info to determine is DS8k lun0
> ```

fix text "determine if its DS8k lun0"

---

*[local/scbe/scbe.go, line 389 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq0x4S5wConVcafVE2:-LFq0x4S5wConVcafVE3:b-1knaz0) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/local/scbe/scbe.go#L389)):*
> ```Go
> 
> 	// Mounter mount will use this extra info to determine is DS8k lun0
> 	lunNumber, ok := lunNumberDict[scbeVolume.WWN]
> ```

Sorry I don't understand what is it?
remember ubiquity server is thread save.
the get and set should be atomic

and what happened if ubiquity crash? is this data should be consistent?

Can we get the lun number from SCBE? volume detail? instead of save it in a memory map.

---

*[local/scbe/scbe.go, line 394 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq2phi6TJddPvwUcMM:-LFq2phi6TJddPvwUcMN:b-5i72xh) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/local/scbe/scbe.go#L394)):*
> ```Go
> 	}
> 
> 	volConfig[resources.LunNumber] = lunNumber
> ```

but if !ok, then what is the lunNumber?

---

*[local/scbe/scbe.go, line 490 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq2X8GDSGnj2u32Ay8:-LFq2X8H6H7bpfR7uExd:bp679c6) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/local/scbe/scbe.go#L490)):*
> ```Go
> 
> 	//Delete the lun number message in the map, will do nothing if wwn not exist in the map
> 	delete(lunNumberDict, existingVolume.WWN)
> ```

whats happened if it fail?

---

*[remote/mounter/scbe.go, line 46 at r3](https://reviewable.io/reviews/ibm/ubiquity/204#-LFkYUhz-ddJukfb_IMY:-LFq3knZ2EPL94shzVf9:bb1te0n) ([raw file](https://github.com/ibm/ubiquity/blob/e87f9ea7ff6e2a0666c2393aacb0d71bb0d52a99/remote/mounter/scbe.go#L46)):*
<details><summary><i>Previously, hfeish (Fei Huang) wrote…</i></summary><blockquote>

Run with gofmt
</blockquote></details>

please keep standard, call it NewScbeMounterWithMounterUtil()

and then reuse code and update NewScbeMounter to use this function  and pass the blockDeviceMounterUtils := block_device_mounter_utils.NewBlockDeviceMounterUtils() to it
reuse code

---

*[remote/mounter/scbe.go, line 68 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq5GIGATTXlfO-gn3-:-LFq5GIGATTXlfO-gn30:b6amxsf) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/remote/mounter/scbe.go#L68)):*
> ```Go
> 		// For DS8k, it only support FC and without LUN0 mapped default, so we need to do "rescan-scsi-bus.sh -a" to discover the LUN0
> 		// Url: UB-1103 in https://www.ibm.com/support/knowledgecenter/SS6JWS_3.4.0/RN/sc_rn_knownissues.html
> 		s.logger.Debug("volumeConfig: ", logs.Args{{"volumeConfig: ", mountRequest.VolumeConfig}})
> ```

please be more descriptive in the log, what you want to state here?

---

*[remote/mounter/scbe.go, line 69 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq6LIi14o5IYhW6R-S:-LFq6LIi14o5IYhW6R-T:bpstfc6) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/remote/mounter/scbe.go#L69)):*
> ```Go
> 		// Url: UB-1103 in https://www.ibm.com/support/knowledgecenter/SS6JWS_3.4.0/RN/sc_rn_knownissues.html
> 		s.logger.Debug("volumeConfig: ", logs.Args{{"volumeConfig: ", mountRequest.VolumeConfig}})
> 		storageType, ok := mountRequest.VolumeConfig[resources.StorageType]
> ```

1. move all this code to function that called isDS8kLun0(volumeconfig object) - cleaner code
2. what is ubiqutiy crash  during mount, what about does it still work in the second run?  idempotent must be consider here.

---

*[remote/mounter/block_device_mounter_utils/block_device_mounter_utils.go, line 200 at r3](https://reviewable.io/reviews/ibm/ubiquity/204#-LFk_N8WBnXLsyn0XxYc:-LFqA5uR0VDcjRgOIakP:b-rajtdj) ([raw file](https://github.com/ibm/ubiquity/blob/e87f9ea7ff6e2a0666c2393aacb0d71bb0d52a99/remote/mounter/block_device_mounter_utils/block_device_mounter_utils.go#L200)):*
<details><summary><i>Previously, hfeish (Fei Huang) wrote…</i></summary><blockquote>

Check that all the para here not use constant, so keep constent
</blockquote></details>

so just move to const - its the right thing todo anyway.

---

*[remote/mounter/block_device_mounter_utils/block_device_mounter_utils.go, line 174 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq8vJFB1uDKnlreivD:-LFq8vJG6Ewt6XUhkef4:b1ge9t5) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/remote/mounter/block_device_mounter_utils/block_device_mounter_utils.go#L174)):*
> ```Go
> 		}
> 	}
> 	if err := b.blockDeviceUtils.Rescan(block_device_utils.BOTH); err != nil {
> ```

what is BOTH?  its confusing

---

*[remote/mounter/block_device_mounter_utils/block_device_mounter_utils.go, line 189 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq9t0_0d1kzh-cdP6q:-LFq9t0aBI0ofPZ2cgxH:b-slm872) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/remote/mounter/block_device_mounter_utils/block_device_mounter_utils.go#L189)):*
> ```Go
> // 2. multipathing rescan
> // return error if one of the steps fail
> func (b *blockDeviceMounterUtils) RescanAllTargets(withISCSI bool, wwn string) error {
> ```

why you are not using the same RescanAll but just with param for all targets?  instead of duplicate this rescan logic

---

*[remote/mounter/block_device_utils/block_device_utils_test.go, line 82 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFqAdepA8R_h5lmRe0o:-LFqAdepA8R_h5lmRe0p:bi5eci4) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/remote/mounter/block_device_utils/block_device_utils_test.go#L82)):*
> ```Go
> 		It("Rescan SCSI fails if rescan-scsi-bus command missing", func() {
> 			fakeExec.IsExecutableReturns(cmdErr)
> 			err = bdUtils.Rescan(block_device_utils.BOTH)
> ```

i don't like you changed this BOTH thing its not clear and made you change a lot of things.

---

*[remote/mounter/block_device_utils/rescan.go, line 28 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq8WNiAo_W_quKPxDH:-LFq8WNiAo_W_quKPxDI:bdt57cm) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/remote/mounter/block_device_utils/rescan.go#L28)):*
> ```Go
> 
> 	switch protocol {
> 	case BOTH:
> ```

this is very confusing
what is BOTH mean?    please be more descriptive here.

---

*[remote/mounter/block_device_utils/rescan.go, line 52 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFqBNLw36SErD-LyOmi:-LFqBNLx8ir0Ocdjij18:b1silsn) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/remote/mounter/block_device_utils/rescan.go#L52)):*
> ```Go
> }
> 
> func (b *blockDeviceUtils) RescanSCSI(str string) error {
> ```

please call it rescanArgs instead of str which is not descriptive.

---

*[resources/resources.go, line 85 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFqBiDrD3o5qtW31Qrx:-LFqBiDrD3o5qtW31Qry:b6jqqp2) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/resources/resources.go#L85)):*
> ```
> const LunNumber = "LunNumber"
> const StorageType = "StorageType"
> ```
what does it mean? is it relevant?  please be more descriptive.

---


*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

shay-berman commented:
@hfeish  I review this PR, and I think you need to do a revisit according to my comment. The basic issue is that you don't need to save lun ID in the ubiquity process memory - it could lead to consistent\concurrency issues.

Please review the comments.

shay-berman commented:
Review status: 3 of 17 files reviewed, 30 unresolved discussions (waiting on @beckmani, @ranhrl, @tzurE, @hfeish, @shay-berman, and @alonm)

---

*[Dockerfile, line 10 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq-enzBExMqiv6VhlU:-LG5UVE08uZ0WfJOyzrk:bfwm2rn) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/Dockerfile#L10)):*
<details><summary><i>Previously, shay-berman wrote…</i></summary><blockquote>

??
how its related to the PR?
</blockquote></details>

please reply Fei

---

*[local/scbe/scbe.go, line 55 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq0lHr4VBUMU9yQeLz:-LG5Uhib3Ttd3Oo3RfiZ:bwpricb) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/local/scbe/scbe.go#L55)):*
<details><summary><i>Previously, shay-berman wrote…</i></summary><blockquote>

what is this?  be careful ubiquity is thread save 
</blockquote></details>

ok thanks for removing it

---

*[local/scbe/scbe.go, line 389 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq0x4S5wConVcafVE2:-LG5V28s0xjgyJOopYVl:b-2mbpjl) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/local/scbe/scbe.go#L389)):*
<details><summary><i>Previously, shay-berman wrote…</i></summary><blockquote>

Sorry I don't understand what is it?
remember ubiquity server is thread save.
the get and set should be atomic

and what happened if ubiquity crash? is this data should be consistent?

Can we get the lun number from SCBE? volume detail? instead of save it in a memory map.
</blockquote></details>

ok thanks for fixing it

---

*[local/scbe/scbe.go, line 394 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq2phi6TJddPvwUcMM:-LG5Uyzf4C6NqztCg3p6:bwpricb) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/local/scbe/scbe.go#L394)):*
<details><summary><i>Previously, shay-berman wrote…</i></summary><blockquote>

but if !ok, then what is the lunNumber?
</blockquote></details>

ok thanks for removing it

---

*[local/scbe/scbe.go, line 386 at r5](https://reviewable.io/reviews/ibm/ubiquity/204#-LG5VUUCC8PAzaFIqoSF:-LG5VUUCC8PAzaFIqoSG:b-frgllo) ([raw file](https://github.com/ibm/ubiquity/blob/f478fcc9cc17e3b78dcee3e143c54036ec6a80b8/local/scbe/scbe.go#L386)):*
> ```Go
> 
> 	volConfig[resources.ScbeKeyVolAttachToHost] = volMapInfo.Host
> 	volConfig[resources.LunNumber] = volMapInfo.LunNumber
> ```

please keep the conversion of keys ScbeKeyVolAttachLunNumToHost

---

*[local/scbe/scbe_rest_client_test.go, line 189 at r5](https://reviewable.io/reviews/ibm/ubiquity/204#-LG5WhuU7vt4Q1lJJUl5:-LG5WhuU7vt4Q1lJJUl6:b-drqewh) ([raw file](https://github.com/ibm/ubiquity/blob/f478fcc9cc17e3b78dcee3e143c54036ec6a80b8/local/scbe/scbe_rest_client_test.go#L189)):*
> ```Go
>             Expect(err).NotTo(HaveOccurred())
>             Expect(volMapInfo.Host).To(Equal(fakeHost))
>             Expect(volMapInfo.LunNumber).To(Equal(fakeLunNumber))
> ```

where is the fakelunnum define? i don't see it in the commit.

---

*[remote/mounter/scbe.go, line 69 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq6LIi14o5IYhW6R-S:-LG5XYUm4kh77lMSSZ-G:bdi3sys) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/remote/mounter/scbe.go#L69)):*
<details><summary><i>Previously, shay-berman wrote…</i></summary><blockquote>

1. move all this code to function that called isDS8kLun0(volumeconfig object) - cleaner code
2. what is ubiqutiy crash  during mount, what about does it still work in the second run?  idempotent must be consider here.
</blockquote></details>

Fei, please reply on my comment so i will know that you fix them. then its easier for me to resolve the comment.

---

*[remote/mounter/scbe.go, line 63 at r5](https://reviewable.io/reviews/ibm/ubiquity/204#-LG5Xm4b1QXDkrl1fi2l:-LG5Xm4b1QXDkrl1fi2m:b6ijk6c) ([raw file](https://github.com/ibm/ubiquity/blob/f478fcc9cc17e3b78dcee3e143c54036ec6a80b8/remote/mounter/scbe.go#L63)):*
> ```Go
> 		// For DS8k, "rescan-scsi-bus.sh -r" cannot discover the LUN0, need to use "rescan-scsi-bus.sh -a" instead
> 		s.logger.Debug("volumeConfig: ", logs.Args{{"volumeConfig: ", mountRequest.VolumeConfig}})
> 		isDs8kLun0 := isDS8kLun0(mountRequest)
> ```

Fie , there is no need to use this isDs8kLun0 variable, just use the function in the if statement.   
you know - "less is more" :-)

---

*[remote/mounter/scbe.go, line 141 at r5](https://reviewable.io/reviews/ibm/ubiquity/204#-LG5Yd9PBBei1hNl73JM:-LG5Yd9PBBei1hNl73JN:b51k9pq) ([raw file](https://github.com/ibm/ubiquity/blob/f478fcc9cc17e3b78dcee3e143c54036ec6a80b8/remote/mounter/scbe.go#L141)):*
> ```Go
> 
> func isDS8kLun0(mountRequest resources.MountRequest) bool {
> 	storageType, ok := mountRequest.VolumeConfig[resources.StorageType]
> ```

please use err instead of ok  (i think its standard way - double check me)

---

*[remote/mounter/block_device_mounter_utils/block_device_mounter_utils.go, line 174 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq8vJFB1uDKnlreivD:-LG5Z-6f2VLWBbs14WWO:byand16) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/remote/mounter/block_device_mounter_utils/block_device_mounter_utils.go#L174)):*
<details><summary><i>Previously, shay-berman wrote…</i></summary><blockquote>

what is BOTH?  its confusing
</blockquote></details>

please reply my comment so i will know what you did to resolve it

---

*[remote/mounter/block_device_utils/block_device_utils_test.go, line 100 at r5](https://reviewable.io/reviews/ibm/ubiquity/204#-LG5ZCM50MH0BWI4p9ka:-LG5ZCM50MH0BWI4p9kb:b1qzlvv) ([raw file](https://github.com/ibm/ubiquity/blob/f478fcc9cc17e3b78dcee3e143c54036ec6a80b8/remote/mounter/block_device_utils/block_device_utils_test.go#L100)):*
> ```Go
> 		})
> 		It("Rescan fails if unknown protocol", func() {
> 			err = bdUtils.Rescan(3)
> ```

what is 3?

---

*[remote/mounter/block_device_utils/rescan.go, line 32 at r5](https://reviewable.io/reviews/ibm/ubiquity/204#-LG5ZGb86Kw7BZC7WPDh:-LG5ZGb86Kw7BZC7WPDi:b-qzisol) ([raw file](https://github.com/ibm/ubiquity/blob/f478fcc9cc17e3b78dcee3e143c54036ec6a80b8/remote/mounter/block_device_utils/rescan.go#L32)):*
> ```Go
> 	case ISCSI:
> 		return b.RescanISCSI()
> 	case FC:
> ```

why u call it FC?    
its confusing
i don't think u need special protocol for that, it sound more like argument for the rescan.

---


*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

shay-berman commented:
Review status: 2 of 17 files reviewed, 27 unresolved discussions (waiting on @beckmani, @ranhrl, @tzurE, @hfeish, @shay-berman, and @alonm)

---

*[Dockerfile, line 10 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq-enzBExMqiv6VhlU:-LGVhc-r0HN965E57aYx:bwqdl1i) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/Dockerfile#L10)):*
<details><summary><i>Previously, hfeish (Fei Huang) wrote…</i></summary><blockquote>

if use openssl=1.0.2n-r0 , it will fail when build the image, so use 1.0.2o instead
</blockquote></details>

this is strange,  please check it out
in the dev its already openssl=1.0.2o-r0

maybe u didn't create this branch from dev!  please check

---

*[remote/mounter/scbe.go, line 146 at r6](https://reviewable.io/reviews/ibm/ubiquity/204#-LGVpTNAChZxPQ82Wuuu:-LGVpTNAChZxPQ82Wuuv:b26p8k6) ([raw file](https://github.com/ibm/ubiquity/blob/cfdd1eea5b48d85345f2a97bec18d368e18c4f1c/remote/mounter/scbe.go#L146)):*
> ```Go
> 	lunNumber, ok := mountRequest.VolumeConfig[resources.ScbeKeyVolAttachLunNumToHost]
> 	if !ok {
> 		lunNumber = -1
> ```

just return here false, instead of setting to -1

---

*[remote/mounter/scbe_test.go, line 73 at r6](https://reviewable.io/reviews/ibm/ubiquity/204#-LGVpeNT4oFDyhu-kJY1:-LGVpeNT4oFDyhu-kJY2:ba8hn57) ([raw file](https://github.com/ibm/ubiquity/blob/cfdd1eea5b48d85345f2a97bec18d368e18c4f1c/remote/mounter/scbe_test.go#L73)):*
> ```Go
> 		})
> 
> 		It("should fail to discover svc with lun1 if failed to discover with '-r' ", func() {
> ```

please make sure u have enough testing and full coverage for the change.
its hard for me to see the full coverage

---

*[remote/mounter/block_device_mounter_utils/block_device_mounter_utils.go, line 179 at r6](https://reviewable.io/reviews/ibm/ubiquity/204#-LGVqEkJEtD_xxqPWlL0:-LGVqEkK1w4vuvS_fQ77:b-i8dwba) ([raw file](https://github.com/ibm/ubiquity/blob/cfdd1eea5b48d85345f2a97bec18d368e18c4f1c/remote/mounter/block_device_mounter_utils/block_device_mounter_utils.go#L179)):*
> ```Go
> 		}
> 	} else {
> 		if err := b.blockDeviceUtils.Rescan(block_device_utils.FC); err != nil {
> ```

its confusing
what is FC vs SCSI???    
its not readable

as i mentioned before - you need to pass extra argument for the rescan. so you still have protocol but extra param for rescans.

---


*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

hfeish commented:
Review status: 1 of 16 files reviewed, 27 unresolved discussions (waiting on @beckmani, @ranhrl, @tzurE, @hfeish, @shay-berman, and @alonm)

---

*[Dockerfile, line 10 at r4](https://reviewable.io/reviews/ibm/ubiquity/204#-LFq-enzBExMqiv6VhlU:-LGYCKuLE1a5WEdhbHHS:b-oaiacs) ([raw file](https://github.com/ibm/ubiquity/blob/fda0a7a649c124e61927c5ab99e8ceaa0014a6d0/Dockerfile#L10)):*
<details><summary><i>Previously, shay-berman wrote…</i></summary><blockquote>

this is strange,  please check it out
in the dev its already openssl=1.0.2o-r0

maybe u didn't create this branch from dev!  please check
</blockquote></details>

It because it still 1.0.2n-r0 when I create this branch from dev at Apr 2, too long ago.
SO I change to 1.0.2o-r0 here otherwise it will fail to build the test build

---

*[remote/mounter/scbe.go, line 146 at r6](https://reviewable.io/reviews/ibm/ubiquity/204#-LGVpTNAChZxPQ82Wuuu:-LGYCvOvE8Dlz95hh64c:bb6un1j) ([raw file](https://github.com/ibm/ubiquity/blob/cfdd1eea5b48d85345f2a97bec18d368e18c4f1c/remote/mounter/scbe.go#L146)):*
<details><summary><i>Previously, shay-berman wrote…</i></summary><blockquote>

just return here false, instead of setting to -1
</blockquote></details>

Done

---

*[remote/mounter/scbe_test.go, line 73 at r6](https://reviewable.io/reviews/ibm/ubiquity/204#-LGVpeNT4oFDyhu-kJY1:-LGYD07HBQMCjcDzvVLG:b-3r6yrr) ([raw file](https://github.com/ibm/ubiquity/blob/cfdd1eea5b48d85345f2a97bec18d368e18c4f1c/remote/mounter/scbe_test.go#L73)):*
<details><summary><i>Previously, shay-berman wrote…</i></summary><blockquote>

please make sure u have enough testing and full coverage for the change.
its hard for me to see the full coverage
</blockquote></details>

Add test cases

---

*[remote/mounter/block_device_mounter_utils/block_device_mounter_utils.go, line 179 at r6](https://reviewable.io/reviews/ibm/ubiquity/204#-LGVqEkJEtD_xxqPWlL0:-LGYDDI19bKubj-8kctW:b-x7rtky) ([raw file](https://github.com/ibm/ubiquity/blob/cfdd1eea5b48d85345f2a97bec18d368e18c4f1c/remote/mounter/block_device_mounter_utils/block_device_mounter_utils.go#L179)):*
<details><summary><i>Previously, shay-berman wrote…</i></summary><blockquote>

its confusing
what is FC vs SCSI???    
its not readable

as i mentioned before - you need to pass extra argument for the rescan. so you still have protocol but extra param for rescans.
</blockquote></details>

use readable para here

---


*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity/204)*
<!-- Sent from Reviewable.io -->

yixuan0825 commented:
@shay-berman ， Hi Shay, could you help to review the new patch based on Fei's patch and your comments.

shay-berman commented:
what is the status of this PR?

yixuan0825 commented:
It's the pull request by Fei, and it is the old design. I will close this PR.
